kaldb/src/main/java/com/slack/kaldb/clusterManager/SnapshotDeletionService.java

```diff
@@ -0,0 +1,224 @@
+   * would see that respected throughout the duration of their lifespan.
+   *
+   * <p>Deletion of expired snapshots should gracefully handle the scenario where immediately prior
+   * to reaching the end of the configured snapshot lifespan, an on-demand request is received. In
+   * this scenario a new replica would have been created, and the object would be currently
+   * downloading for servicing the request. The expectation from the end user would be that the
+   * request would still be served, as when the request was made it was still "in compliance" with
+   * the lifespan configuration.
+   *
+   * <p>When creating on-demand replicas, care should be taken to not create replicas for snapshots
+   * that already exceed their maximum lifespan to prevent a race condition between on-demand
+   * creation and snapshot deletion.
+   */
+  protected int deleteExpiredSnapshotsWithoutReplicas() {
+    Timer.Sample deletionTimer = Timer.start(meterRegistry);
+
+    Set<String> snapshotIdsWithReplicas =
+        replicaMetadataStore
+            .getCached()
+            .stream()
+            .map(replicaMetadata -> replicaMetadata.snapshotId)
+            .filter(snapshotId -> snapshotId != null && !snapshotId.isEmpty())
+            .collect(Collectors.toUnmodifiableSet());
+
+    long expirationCutoff =
+        Instant.now()
+            .minus(
+                managerConfig.getSnapshotDeletionServiceConfig().getSnapshotLifespanMins(),
+                ChronoUnit.MINUTES)
+            .toEpochMilli();
+    AtomicInteger successCounter = new AtomicInteger(0);
+    List<ListenableFuture<?>> deletedSnapshotList =
+        snapshotMetadataStore
+            .getCached()
+            .stream()
+            // only snapshots that only contain data prior to our cutoff, and have no replicas
+            .filter(
+                snapshotMetadata ->
+                    snapshotMetadata.endTimeEpochMs < expirationCutoff
+                        && !snapshotIdsWithReplicas.contains(snapshotMetadata.name))
+            .map(
+                snapshotMetadata -> {
+                  ListenableFuture<?> future =
+                      Futures.submit(
+                          () -> {
+                            // These futures are rate-limited so that we can more evenly distribute
+                            // the load to the downstream services (metadata, s3). There is no
+                            // urgency to complete the deletes, so limiting the maximum rate allows
+                            // us to avoid unnecessary spikes.
+                            rateLimiter.acquire();
+
+                            // We first delete the snapshot reference - if for some reason we fail
+                            // to delete the object from the object storage, the orphaned object
+                            // cleanup service will attempt to delete it later
+                            snapshotMetadataStore.deleteSync(snapshotMetadata);
```

Comment:
Let's delete the snapshot metadata after deleting the data on S3 so we can reduce the need for object clean up. That way on the next run, we can delete the missed files.

Reply:
Done.